File: MonteCarlo.py
from numpy import *
import random
from math import sqrt

# TODO: Naming rule is bad. Not very software engineering friendly.

# Parameter settings
Sample_Number = 10000
T = 20                          # trading in year
dT = 1 / 250                    # in year (250 trading day per year)
Step_Num = int(T / dT)
Inflation = 3.5 / 100           # inflation rate
Initial_Investment = 100000

mu_Aggressive = 9.4324 / 100
sigma_Aggressive = 15.675 / 100

mu_Conservative = 6.189 / 100
sigma_Conservative = 6.3438 / 100

median = int(Sample_Number * 0.5)
top10 = int(Sample_Number * (1 - 0.1))
bottom10 = int(Sample_Number * (1 -0.9))

# Construct matrix as sample paths
dZ_Aggressive = array([random.gauss(0, 1) for i in range(Sample_Number * Step_Num)]) # Gauss Distribution
dZ_Aggressive.shape = (Sample_Number, Step_Num)

aggressiveFlow = zeros((Sample_Number, Step_Num))
aggressiveFlow[:, 0] = Initial_Investment

for i in range(Step_Num - 1):
    aggressiveFlow[:, i + 1] = aggressiveFlow[:, i] * exp(
        (mu_Aggressive - 0.5 * sigma_Aggressive**2 - Inflation) * dT +
        sigma_Aggressive * sqrt(dT) * dZ_Aggressive[:, i]
    )                   # Weiner Process

# TODO: the generation of sample paths can be generalized. Refactor code is necessary in future.
dZ_Conservative = array([random.gauss(0, 1) for i in range(Sample_Number * Step_Num)]) # Gauss Distribution
dZ_Conservative.shape = (Sample_Number, Step_Num)

conservativeFlow = zeros((Sample_Number, Step_Num))
conservativeFlow[:, 0] = Initial_Investment

for i in range(Step_Num - 1):
    conservativeFlow[:, i + 1] = conservativeFlow[:, i] * exp(
        (mu_Conservative - 0.5 * sigma_Conservative**2 - Inflation) * dT +
        sigma_Conservative * sqrt(dT) * dZ_Conservative[:, i]
    )                   # Weiner Process


# Output result
resultAggressive = aggressiveFlow[:, Step_Num - 1]
resultAggressive.sort()
print("Aggressive 20 year median is: ", resultAggressive[median], end='\n')
print("Aggressive 20 year top 10% is: ", resultAggressive[top10], end='\n')
print("Aggressive 20 year bottom 10% is: ", resultAggressive[bottom10], end='\n')

resultConservative = conservativeFlow[:, Step_Num - 1]
resultConservative.sort()
print("Conservative 20 year median is: ", resultConservative[median], end='\n')
print("Conservative 20 year top 10% is: ", resultConservative[top10], end='\n')
print("Conservative 20 year bottom 10% is: ", resultConservative[bottom10], end='\n')

File: MonteCarlo.java
import java.util.*;
import java.lang.Math;

public class MonteCarlo {
    // write this program in a "macro" way in main().
    private static final int SAMPLE_NUM = 10000;
    private static final int TIME = 20;                          // time in years
    private static final double DELTA_TIME_INTERVAL = 1.0 / 250;   // around 250 trading days per year

    private static final double INFLATION = 3.5 / 100;
    private static final int INITIAL_INVESTMENT = 100000;
    private static final double MU_AGGRESSIVE = 9.4324 / 100;
    private static final double SIGMA_AGGRESSIVE = 15.675 / 100;
    private static final double MU_CONSERVATIVE = 6.189 / 100;
    private static final double SIGMA_CONSERVATIVE = 6.3438 / 100;

    public static void main(String[] args) {
        int steps = (int) (TIME / DELTA_TIME_INTERVAL);
        int median = (int) (SAMPLE_NUM * 0.5);
        int top10 = (int) (SAMPLE_NUM * (1 - 0.1));
        int bottom10 = (int) (SAMPLE_NUM * (1 - 0.9));

        double[][] value_aggressive = new double[SAMPLE_NUM][2];
        double[][] value_conservative = new double[SAMPLE_NUM][2];

        for (int i = 0; i < SAMPLE_NUM; i++) {
            value_aggressive[i][0] = INITIAL_INVESTMENT;
            value_conservative[i][0] = INITIAL_INVESTMENT;
        }

        weinerProcess(value_aggressive, steps, MU_AGGRESSIVE, SIGMA_AGGRESSIVE);
        weinerProcess(value_conservative, steps, MU_CONSERVATIVE, SIGMA_CONSERVATIVE);

        double[] output_aggressive = new double[SAMPLE_NUM];
        double[] output_conservative = new double[SAMPLE_NUM];

        for (int i = 0; i < SAMPLE_NUM; i++) {
            output_aggressive[i] = value_aggressive[i][1];
            output_conservative[i] = value_conservative[i][1];
        }

        Arrays.sort(output_aggressive);
        Arrays.sort(output_conservative);

        System.out.println("Top 10% return of aggressive portfolio is: " + output_aggressive[top10]);
        System.out.println("Median return of aggressive portfolio is: " + output_aggressive[median]);
        System.out.println("Bottom 10% return of aggressive portfolio is: " + output_aggressive[bottom10]);
        System.out.println();

        System.out.println("Top 10% return of conservative portfolio is: " + output_conservative[top10]);
        System.out.println("Median return of conservative portfolio is: " + output_conservative[median]);
        System.out.println("Bottom 10% return of conservative portfolio is: " + output_conservative[bottom10]);

    }

    private static void weinerProcess(double[][] valueMatrix, int steps, double mu, double sigma) {
        Random random = new Random();
        int curStep = 0;
        int nextStep = 1 - curStep;
        for (int i = 1; i < steps; i++) {
            for (int j = 0; j < SAMPLE_NUM; j++) {
                valueMatrix[j][nextStep] = valueMatrix[j][curStep] *
                        Math.exp( (mu - 0.5 * sigma * sigma - INFLATION) * DELTA_TIME_INTERVAL +
                                sigma * Math.sqrt(DELTA_TIME_INTERVAL) * random.nextGaussian()
                        );
            }
            curStep = nextStep;
            nextStep = 1 - curStep;
        }
        if (nextStep < curStep) { // switch result if final value is in the [0] column
            for (int i = 0; i < SAMPLE_NUM; i++) {
                double tmp = valueMatrix[i][nextStep];
                valueMatrix[i][nextStep] = valueMatrix[i][curStep];
                valueMatrix[i][curStep] = tmp;
            }
        }
    }
}
